src/main/java/winstone/Launcher.java

```diff
@@ -229,9 +231,19 @@ private void writePortToFileIfNeeded() throws IOException {
                     Path portFile = Paths.get(portFileName);
                     Path portDir = portFile.getParent();
                     Files.createDirectories(portDir);
-                    try (BufferedWriter writer = Files.newBufferedWriter(portFile, StandardCharsets.UTF_8)) {
+                    Path tmpPath = Files.createTempFile(portDir, "temp", "");
```

Comment:
Using `portDir` here is great! That ensures that the temporary file is created on the same file system as the target, so the atomic move is more likely to succeed. :thumbsup:

The use of `temp` as the prefix is a bit generic: why not use the name of the destination file as the prefix? That way the temporary file is more obviously connected to its destination which can make debugging easier.

Similarly, if you don't have a particular suffix in mind, I think it is more idiomatic to use `null` as the suffix, which gives us the default suffix of `.tmp`. I think it is better to stick with the default if we don't have a specific customization in mind.

```suggestion
                    Path tmpPath = Files.createTempFile(portDir, portFile.getFileName().toString(), null);
```